NuGet.Build.Tasks caches CredentialProvider device flow timeout.

### NuGet Product Used

MSBuild.exe

### Product Version

Nuget 6.5.0.149 (does not seems to be fixed in recent code yet)

### Worked before?

unknown

### Impact

It's more difficult to complete my work

### Repro Steps & Context

We are working on MSBuild server which is concept of long running process which will execute consequent builds so it can effectively reuse memory caches between builds.  People were reporting intermittent inability to use interactive restore.

### Analysis
Root cause: Task caches result of timeouted CredentialProvider device flow. Consequent runs do not call CredentialProvider and use cached failure value.

Cache is rooted by: https://github.com/NuGet/NuGet.Client/blob/3a0c9a47e24642bb4a4465dec7b4d1c379367855/src/NuGet.Core/NuGet.Protocol/HttpSource/HttpHandlerResourceV3.cs#L40

Result is "wrongly" identified as cacheable at line nr 125 bellow:
https://github.com/NuGet/NuGet.Client/blob/3a0c9a47e24642bb4a4465dec7b4d1c379367855/src/NuGet.Core/NuGet.Credentials/CredentialService.cs#L106-L129

There also seems to to be related bug in CredetionProvider which causes null exception as can be seen in log bellow.
https://github.com/microsoft/artifacts-credprovider/blob/e19f61d6c0697a73fb6b08cd3c9c687f9aa97d36/CredentialProvider.Microsoft/CredentialProviders/Vsts/AdalTokenProvider.cs#L46-L59
Constructor on line 59 fails on null exception when it timeouted because it will pass in null as argument.

### Repro steps
It is hard to repro with existing MSBuild as it has some workarounds around this bug.
I do not believe repro is necessary, log and analysis seem to be enough to understand root cause of this bug.
However if I would need to repro it I would:

- prepare project which needs interactive restore
- create console app and link it with msbuild.dll
- in that console app hardcore it to build given project twice
- let first project build timeout on sing in requests
- observe second build to fail without sign in prompt

If repro is really needed please ping me and I can help with building it or testing it.

### Verbose Logs

```shell
Bellow is related part from msbuild binlog. I believe it is giving us enough context and it validates above analysis:

RestoreTask
    Assembly = E:\dev\msbuild\artifacts\bin\bootstrap\net472\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.Build.Tasks.dll
    Parameters
        RestoreRecursive = True
        HideWarningsAndErrors = False
        Interactive = True
        RestoreGraphItems
            89980f99-15df-4772-8794-ae61c4040b7c
                MSBuildSourceProjectFile = E:\dev-tmp\testme\testme.csproj
                ProjectUniqueName = E:\dev-tmp\testme\testme.csproj
                Type = RestoreSpec
                MSBuildSourceTargetName = _IsProjectRestoreSupported
                OriginalItemSpec = E:\dev-tmp\testme\testme.csproj
            a7b10e3c-a242-45e1-9a8a-7cd9788b5806
                SkipContentFileWrite = 
                MSBuildSourceProjectFile = E:\dev-tmp\testme\testme.csproj
                FallbackFolders = C:\Program Files (x86)\Microsoft Visual Studio\Shared\NuGetPackages;C:\Program Files\dotnet\sdk\NuGetFallbackFolder
                NuGetLockFilePath = 
                ProjectUniqueName = E:\dev-tmp\testme\testme.csproj
                RestoreLockedMode = 
                Sources = https://api.nuget.org/v3/index.json;https://pkgs.dev.azure.com/dnceng/internal/_packaging/7.0.100-rc.1.22429.3-shipping/nuget/v3/index.json;C:\Program Files\dotnet\library-packs
                Type = ProjectSpec
                Version = 1.0.0
                _CentralPackageVersionsEnabled = 
                ValidateRuntimeAssets = false
                CrossTargeting = 
                WarningsAsErrors = ;NU1605;SYSLIB0011
                RestorePackagesWithLockFile = 
                RestoreLegacyPackagesDirectory = 
                ProjectStyle = PackageReference
                RuntimeSupports = 
                TreatWarningsAsErrors = false
                RuntimeIdentifiers = ;win-x64
                WarningsNotAsErrors = 
                NoWarn = 1701;1702;IL2121
                CentralPackageTransitivePinningEnabled = 
                OutputPath = E:\dev-tmp\testme\obj\
                MSBuildSourceTargetName = _IsProjectRestoreSupported
                ProjectPath = E:\dev-tmp\testme\testme.csproj
                CentralPackageVersionOverrideEnabled = 
                ProjectName = testme
                PackagesPath = E:\NugetCache
                OriginalItemSpec = E:\dev-tmp\testme\testme.csproj
                ConfigFilePaths = E:\dev-tmp\testme\NuGet.Config;C:\Users\rokon\AppData\Roaming\NuGet\NuGet.Config;C:\Program Files (x86)\NuGet\Config\Microsoft.VisualStudio.FallbackLocation.config;C:\Program Files (x86)\NuGet\Config\Microsoft.VisualStudio.Offline.config
            e03d5822-4e87-4b9a-a152-87a2f333388b
                MSBuildSourceProjectFile = E:\dev-tmp\testme\testme.csproj
                ProjectUniqueName = E:\dev-tmp\testme\testme.csproj
                Id = AutoMapper
                Type = Dependency
                TargetFrameworks = net7.0
                MSBuildSourceTargetName = _IsProjectRestoreSupported
                OriginalItemSpec = E:\dev-tmp\testme\testme.csproj
                VersionRange = 12.0.1-internal
            8d872925-d5bd-4302-98b5-efc38531415c
                MSBuildSourceProjectFile = E:\dev-tmp\testme\testme.csproj
                ProjectUniqueName = E:\dev-tmp\testme\testme.csproj
                Id = Microsoft.NETCore.App.Runtime.win-x64
                Type = DownloadDependency
                TargetFrameworks = net7.0
                MSBuildSourceTargetName = _IsProjectRestoreSupported
                OriginalItemSpec = E:\dev-tmp\testme\testme.csproj
                VersionRange = [7.0.4]
            a581f697-aa82-448b-a2cb-0f01094a052b
                MSBuildSourceProjectFile = E:\dev-tmp\testme\testme.csproj
                ProjectUniqueName = E:\dev-tmp\testme\testme.csproj
                Id = Microsoft.WindowsDesktop.App.Runtime.win-x64
                Type = DownloadDependency
                TargetFrameworks = net7.0
                MSBuildSourceTargetName = _IsProjectRestoreSupported
                OriginalItemSpec = E:\dev-tmp\testme\testme.csproj
                VersionRange = [7.0.4]
            91bd029f-4c9f-4b40-b3d6-932225623961
                MSBuildSourceProjectFile = E:\dev-tmp\testme\testme.csproj
                ProjectUniqueName = E:\dev-tmp\testme\testme.csproj
                Id = Microsoft.AspNetCore.App.Runtime.win-x64
                Type = DownloadDependency
                TargetFrameworks = net7.0
                MSBuildSourceTargetName = _IsProjectRestoreSupported
                OriginalItemSpec = E:\dev-tmp\testme\testme.csproj
                VersionRange = [7.0.4]
            d9b10af3-489e-4d5a-a97c-3630d09daf6b
                MSBuildSourceProjectFile = E:\dev-tmp\testme\testme.csproj
                ProjectUniqueName = E:\dev-tmp\testme\testme.csproj
                Id = Microsoft.NETCore.App
                Type = FrameworkReference
                TargetFrameworks = net7.0
                PrivateAssets = All
                MSBuildSourceTargetName = _IsProjectRestoreSupported
                OriginalItemSpec = E:\dev-tmp\testme\testme.csproj
            915d08f7-e433-42e0-a656-3bb8c6ac8301
                MSBuildSourceProjectFile = E:\dev-tmp\testme\testme.csproj
                TargetFrameworkProfile = 
                TargetFramework = net7.0
                TargetFrameworkIdentifier = .NETCoreApp
                ProjectUniqueName = E:\dev-tmp\testme\testme.csproj
                AssetTargetFallback = ;net461;net462;net47;net471;net472;net48;net481
                PackageTargetFallback = 
                Type = TargetFrameworkInformation
                TargetPlatformVersion = 
                TargetPlatformMoniker = 
                TargetFrameworkMoniker = .NETCoreApp,Version=v7.0
                MSBuildSourceTargetName = _IsProjectRestoreSupported
                RuntimeIdentifierGraphPath = C:\Program Files\dotnet\sdk\7.0.202\RuntimeIdentifierGraph.json
                CLRSupport = 
                TargetPlatformIdentifier = 
                TargetFrameworkVersion = v7.0
                TargetPlatformMinVersion = 
                OriginalItemSpec = E:\dev-tmp\testme\testme.csproj
    (in) RestoreGraphItems Count '8'
    (in) RestoreDisableParallel 'False'
    (in) RestoreNoCache 'False'
    (in) RestoreIgnoreFailedSources 'False'
    (in) RestoreRecursive 'True'
    (in) RestoreForce 'False'
    (in) HideWarningsAndErrors 'False'
    (in) RestoreForceEvaluate 'False'
    (in) RestorePackagesConfig 'False'
    Assembly loaded during TaskRun (NuGet.Build.Tasks.RestoreTask): NuGet.Packaging, Version=6.5.0.149, Culture=neutral, PublicKeyToken=31bf3856ad364e35 (location: E:\dev\msbuild\artifacts\bin\bootstrap\net472\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.Packaging.dll, MVID: 4682e763-f450-447f-93b4-5dfc8ab16e13, AppDomain: [Default])
    Assembly loaded during TaskRun (NuGet.Build.Tasks.RestoreTask): NuGet.LibraryModel, Version=6.5.0.149, Culture=neutral, PublicKeyToken=31bf3856ad364e35 (location: E:\dev\msbuild\artifacts\bin\bootstrap\net472\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.LibraryModel.dll, MVID: bbcf1ffb-8b16-4c3c-a4db-e6730ea0b62d, AppDomain: [Default])
    Assembly loaded during TaskRun (NuGet.Build.Tasks.RestoreTask): NuGet.Protocol, Version=6.5.0.149, Culture=neutral, PublicKeyToken=31bf3856ad364e35 (location: E:\dev\msbuild\artifacts\bin\bootstrap\net472\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.Protocol.dll, MVID: 942c272f-5083-46e2-b29d-da0f828a8326, AppDomain: [Default])
    Assembly loaded during TaskRun (NuGet.Build.Tasks.RestoreTask): NuGet.Credentials, Version=6.5.0.149, Culture=neutral, PublicKeyToken=31bf3856ad364e35 (location: E:\dev\msbuild\artifacts\bin\bootstrap\net472\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.Credentials.dll, MVID: 6953f29b-c883-4d0c-b809-46e2cc1d14f0, AppDomain: [Default])
    Assembly loaded during TaskRun (NuGet.Build.Tasks.RestoreTask): NuGet.PackageManagement, Version=6.5.0.149, Culture=neutral, PublicKeyToken=31bf3856ad364e35 (location: E:\dev\msbuild\artifacts\bin\bootstrap\net472\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.PackageManagement.dll, MVID: a090fa76-1ca0-4832-bd5a-af4375e13f8f, AppDomain: [Default])
    X.509 certificate chain validation will use the default trust store selected by .NET.
    X.509 certificate chain validation will use the default trust store selected by .NET.
    Assembly loaded during TaskRun (NuGet.Build.Tasks.RestoreTask): NuGet.DependencyResolver.Core, Version=6.5.0.149, Culture=neutral, PublicKeyToken=31bf3856ad364e35 (location: E:\dev\msbuild\artifacts\bin\bootstrap\net472\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.DependencyResolver.Core.dll, MVID: 4741d55d-6810-4900-ac9d-f4cce6ed0c09, AppDomain: [Default])
    Running restore with 24 concurrent jobs.
    Reading project file
        Reading project file E:\dev-tmp\testme\testme.csproj.
    Assembly loaded during TaskRun (NuGet.Build.Tasks.RestoreTask): System.Runtime.Serialization, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089 (location: C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\System.Runtime.Serialization\v4.0_4.0.0.0__b77a5c561934e089\System.Runtime.Serialization.dll, MVID: e0272679-dd10-40b0-adef-29cb5a2d4ab0, AppDomain: [Default])
    The restore inputs for 'testme' have changed. Continuing restore.
    Restoring packages for
        Restoring packages for E:\dev-tmp\testme\testme.csproj...
        Restoring packages for .NETCoreApp,Version=v7.0...
    Assembly loaded during TaskRun (NuGet.Build.Tasks.RestoreTask): System.Net.Http, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a (location: C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\System.Net.Http\v4.0_4.0.0.0__b03f5f7f11d50a3a\System.Net.Http.dll, MVID: 16bc693b-b84a-493b-8a40-c4959d218321, AppDomain: [Default])
    Assembly loaded during TaskRun (NuGet.Build.Tasks.RestoreTask): System.Data, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089 (location: C:\WINDOWS\Microsoft.Net\assembly\GAC_64\System.Data\v4.0_4.0.0.0__b77a5c561934e089\System.Data.dll, MVID: 46b0a213-4517-4c10-8daa-3339b363ef3b, AppDomain: [Default])
    GET
          GET https://api.nuget.org/v3-flatcontainer/automapper/index.json
    Using C:\Users\rokon\.nuget\plugins\netfx\CredentialProvider.Microsoft\CredentialProvider.Microsoft.exe as a credential provider plugin.
    OK
          OK https://api.nuget.org/v3-flatcontainer/automapper/index.json 126ms
        [CredentialProvider.103634]Running in plug-in mode
        [CredentialProvider.103634]Command-line v0.1.25+cb125622a8f121aadf3e82f315b06fdb42198585: "C:\Users\rokon\.nuget\plugins\netfx\CredentialProvider.Microsoft\CredentialProvider.Microsoft.exe" -Plugin
        [CredentialProvider.103635]Handling 'Request' 'Initialize'. Time elapsed in ms: 1 - Payload: {"ClientVersion":"6.5.0","Culture":"en-US","RequestTimeout":"00:00:30"}
        [CredentialProvider.103635]Sending response: 'Request' 'Initialize'. Time elapsed in ms: 1
        [CredentialProvider.103635]Time elapsed in milliseconds after sending response 'Request' 'Initialize': 1
        [CredentialProvider.103635]Handling 'Request' 'GetOperationClaims'. Time elapsed in ms: 0 - Payload: {}
        [CredentialProvider.103635]Sending response: 'Request' 'GetOperationClaims'. Time elapsed in ms: 4
        [CredentialProvider.103635]Time elapsed in milliseconds after sending response 'Request' 'GetOperationClaims': 6
        [CredentialProvider.103635]Handling 'Request' 'SetLogLevel'. Time elapsed in ms: 0 - Payload: {"LogLevel":"Debug"}
        [CredentialProvider]Sending response: 'Request' 'SetLogLevel'. Time elapsed in ms: 1
        [CredentialProvider]Time elapsed in milliseconds after sending response 'Request' 'SetLogLevel': 3
        [CredentialProvider]Handling 'Request' 'GetAuthenticationCredentials'. Time elapsed in ms: 1 - Payload: {"Uri":"https://pkgs.dev.azure.com/dnceng/internal/_packaging/7.0.100-rc.1.22429.3-shipping/nuget/v3/index.json","IsRetry":false,"IsNonInteractive":false,"CanShowDialog":false}
        [CredentialProvider]Creating a progress reporter with interval: 00:00:02
        [CredentialProvider]Handling auth request, Uri: https://pkgs.dev.azure.com/dnceng/internal/_packaging/7.0.100-rc.1.22429.3-shipping/nuget/v3/index.json, IsRetry: False, IsNonInteractive: False, CanShowDialog: False
        [CredentialProvider]URI: https://pkgs.dev.azure.com/dnceng/internal/_packaging/7.0.100-rc.1.22429.3-shipping/nuget/v3/index.json
        [CredentialProvider]VstsBuildTaskServiceEndpointCredentialProvider - This credential provider must be run under the Team Build tasks for NuGet with external endpoint credentials. Appropriate environment variable needs to be set.
        [CredentialProvider]Skipping NuGetCredentialProvider.CredentialProviders.VstsBuildTaskServiceEndpoint.VstsBuildTaskServiceEndpointCredentialProvider, cannot provide credentials for https://pkgs.dev.azure.com/dnceng/internal/_packaging/7.0.100-rc.1.22429.3-shipping/nuget/v3/index.json
        [CredentialProvider]VstsBuildTaskCredentialProvider - This credential provider must be run under the Team Build tasks for NuGet. Appropriate environment variables must be set.
        [CredentialProvider]Skipping NuGetCredentialProvider.CredentialProviders.VstsBuildTask.VstsBuildTaskCredentialProvider, cannot provide credentials for https://pkgs.dev.azure.com/dnceng/internal/_packaging/7.0.100-rc.1.22429.3-shipping/nuget/v3/index.json
        [CredentialProvider]GET https://pkgs.dev.azure.com/dnceng/internal/_packaging/7.0.100-rc.1.22429.3-shipping/nuget/v3/index.json
        [CredentialProvider]VstsCredentialProvider - Detected a hosted Azure DevOps Service.
        [CredentialProvider]Using NuGetCredentialProvider.CredentialProviders.Vsts.VstsCredentialProvider to try to get credentials for https://pkgs.dev.azure.com/dnceng/internal/_packaging/7.0.100-rc.1.22429.3-shipping/nuget/v3/index.json.
        [CredentialProvider]IsRetry: False
        [CredentialProvider]Could not find cached SessionToken for https://pkgs.dev.azure.com/dnceng/internal/_packaging/7.0.100-rc.1.22429.3-shipping/nuget/v3/index.json
        [CredentialProvider]GET https://pkgs.dev.azure.com/dnceng/internal/_packaging/7.0.100-rc.1.22429.3-shipping/nuget/v3/index.json
        [CredentialProvider]Found AAD Authority from 401 headers: https://login.windows.net/72f988bf-86f1-41af-91ab-2d7cd011db47
        [CredentialProvider]VstsCredentialProvider - Using AAD authority: https://login.windows.net/72f988bf-86f1-41af-91ab-2d7cd011db47
        [CredentialProvider]VstsCredentialProvider - Attempting to acquire bearer token using provider 'ADAL Cache'
        [CredentialProvider]VstsCredentialProvider - Bearer token provider 'ADAL Cache' didn't acquire a token
        [CredentialProvider]VstsCredentialProvider - Attempting to acquire bearer token using provider 'ADAL Windows Integrated Authentication'
        [CredentialProvider]VstsCredentialProvider - Bearer token provider 'ADAL Windows Integrated Authentication' didn't acquire a token
        [CredentialProvider]VstsCredentialProvider - Not running bearer token provider 'ADAL UI'
        [CredentialProvider]VstsCredentialProvider - Attempting to acquire bearer token using provider 'ADAL Device Code'
        [CredentialProvider]DeviceFlow: https://pkgs.dev.azure.com/dnceng/internal/_packaging/7.0.100-rc.1.22429.3-shipping/nuget/v3/index.json
        [CredentialProvider]ATTENTION: User interaction required. 

    **********************************************************************

    To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code C2YA83482 to authenticate.

    **********************************************************************

    Errors
        E:\dev\msbuild\artifacts\bin\bootstrap\net472\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.targets(132,5):     [CredentialProvider]Device flow authentication failed. User was presented with device flow, but didn't react within 90 seconds. [E:\dev-tmp\testme\testme.csproj]
        E:\dev-tmp\testme\testme.csproj error NU1301: Unable to load the service index for source https://pkgs.dev.azure.com/dnceng/internal/_packaging/7.0.100-rc.1.22429.3-shipping/nuget/v3/index.json. [E:\dev-tmp\testme\testme.csproj]
        [CredentialProvider]VstsCredentialProvider - Bearer token provider 'ADAL Device Code' failed with exception:\nSystem.NullReferenceException: Object reference not set to an instance of an object.
   at NuGetCredentialProvider.CredentialProviders.Vsts.AdalTokenProvider.<AcquireTokenWithDeviceFlowAsync>d__6.MoveNext() in E:\A\_work\998\s\CredentialProvider.Microsoft\CredentialProviders\Vsts\AdalTokenProvider.cs:line 59
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGetCredentialProvider.CredentialProviders.Vsts.DeviceCodeFlowBearerTokenProvider.<GetTokenAsync>d__9.MoveNext() in E:\A\_work\998\s\CredentialProvider.Microsoft\CredentialProviders\Vsts\BearerTokenProviders.cs:line 122
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at NuGetCredentialProvider.CredentialProviders.Vsts.VstsCredentialProvider.<HandleRequestAsync>d__8.MoveNext() in E:\A\_work\998\s\CredentialProvider.Microsoft\CredentialProviders\Vsts\VstsCredentialProvider.cs:line 117
        [CredentialProvider]VstsCredentialProvider - Could not obtain credentials for https://pkgs.dev.azure.com/dnceng/internal/_packaging/7.0.100-rc.1.22429.3-shipping/nuget/v3/index.json
        [CredentialProvider]Unable to acquire credentials.
        [CredentialProvider]Sending response: 'Request' 'GetAuthenticationCredentials'. Time elapsed in ms: 91130
        [CredentialProvider]Time elapsed in milliseconds after sending response 'Request' 'GetAuthenticationCredentials': 91131
    Assets file
        Writing assets file to disk. Path: E:\dev-tmp\testme\obj\project.assets.json
    Cache file
        Writing cache file to disk. Path: E:\dev-tmp\testme\obj\project.nuget.cache
    dg file
        Persisting dg to E:\dev-tmp\testme\obj\testme.csproj.nuget.dgspec.json
    Failed to restore E:\dev-tmp\testme\testme.csproj (in 1.65 min).
    NuGet Config files used:
        E:\dev-tmp\testme\NuGet.Config
        C:\Users\rokon\AppData\Roaming\NuGet\NuGet.Config
        C:\Program Files (x86)\NuGet\Config\Microsoft.VisualStudio.FallbackLocation.config
        C:\Program Files (x86)\NuGet\Config\Microsoft.VisualStudio.Offline.config
    Feeds used:
        https://api.nuget.org/v3/index.json
        https://pkgs.dev.azure.com/dnceng/internal/_packaging/7.0.100-rc.1.22429.3-shipping/nuget/v3/index.json
        C:\Program Files\dotnet\library-packs
```

--- a/src/NuGet.Core/NuGet.Credentials/SecurePluginCredentialProvider.cs
+++ b/src/NuGet.Core/NuGet.Credentials/SecurePluginCredentialProvider.cs
@@ -210,6 +210,10 @@ namespace NuGet.Credentials
 
                 taskResponse = new CredentialResponse(result);
             }
+            else if (credentialResponse.ResponseCode == MessageResponseCode.NotFound)
+            {
+                taskResponse = new CredentialResponse(CredentialStatus.UserCanceled);
+            }
             else
             {
                 taskResponse = new CredentialResponse(CredentialStatus.ProviderNotApplicable);

--- a/src/NuGet.Core/NuGet.Protocol/HttpSource/HttpSourceAuthenticationHandler.cs
+++ b/src/NuGet.Core/NuGet.Protocol/HttpSource/HttpSourceAuthenticationHandler.cs
@@ -256,17 +256,11 @@ namespace NuGet.Protocol
                 promptCredentials = await _credentialService
                     .GetCredentialsAsync(_packageSource.SourceUri, proxy, type, message, token);
 
-                if (promptCredentials == null)
-                {
-                    // If this is the case, this means none of the credential providers were able to
-                    // handle the credential request or no credentials were available for the
-                    // endpoint.
-                    authState.Block();
-                }
-                else
-                {
-                    authState.Increment();
-                }
+                // If promptCredentials == null means none of the credential providers were able to
+                // handle the credential request or no credentials were available for the
+                // endpoint, a retry might fix the issue so we increment the authState.
+
+                authState.Increment();
             }
             catch (OperationCanceledException)
             {

--- a/test/NuGet.Core.Tests/NuGet.Credentials.Test/PluginManagerMock.cs
+++ b/test/NuGet.Core.Tests/NuGet.Credentials.Test/PluginManagerMock.cs
@@ -33,6 +33,7 @@ namespace NuGet.Credentials.Test
         public string ProxyPassword { get; }
         public bool PluginLaunched { get; }
         public bool CanShowDialog { get; }
+        public bool MessageCodeNotFound { get; }
 
         internal TestExpectation(
             IEnumerable<OperationClaim> operationClaims,
@@ -45,7 +46,8 @@ namespace NuGet.Credentials.Test
             string proxyUsername = null,
             string proxyPassword = null,
             bool pluginLaunched = true,
-            bool canShowDialog = true)
+            bool canShowDialog = true,
+            bool messageCodeNotFound = false)
         {
             OperationClaims = operationClaims;
             ClientConnectionOptions = connectionOptions;
@@ -58,6 +60,7 @@ namespace NuGet.Credentials.Test
             ProxyUsername = proxyUsername;
             PluginLaunched = pluginLaunched;
             CanShowDialog = canShowDialog;
+            MessageCodeNotFound = messageCodeNotFound;
         }
     }
 
@@ -139,6 +142,21 @@ namespace NuGet.Credentials.Test
                         It.Is<GetAuthenticationCredentialsRequest>(e => e.Uri.Equals(expectations.Uri) && e.CanShowDialog.Equals(expectations.CanShowDialog)),
                         It.IsAny<CancellationToken>()))
                     .ReturnsAsync(new GetAuthenticationCredentialsResponse(expectations.AuthenticationUsername, expectations.AuthenticationPassword, message: null, authenticationTypes: null, responseCode: MessageResponseCode.Success));
+            }
+
+            if (_expectations.MessageCodeNotFound)
+            {
+                _connection.Setup(x => x.SendRequestAndReceiveResponseAsync<SetLogLevelRequest, SetLogLevelResponse>(
+                        It.Is<MessageMethod>(m => m == MessageMethod.SetLogLevel),
+                        It.IsAny<SetLogLevelRequest>(),
+                        It.IsAny<CancellationToken>()))
+                    .ReturnsAsync(new SetLogLevelResponse(MessageResponseCode.Success));
+
+                _connection.Setup(x => x.SendRequestAndReceiveResponseAsync<GetAuthenticationCredentialsRequest, GetAuthenticationCredentialsResponse>(
+                        It.Is<MessageMethod>(m => m == MessageMethod.GetAuthenticationCredentials),
+                        It.Is<GetAuthenticationCredentialsRequest>(e => e.Uri.Equals(expectations.Uri) && e.CanShowDialog.Equals(expectations.CanShowDialog)),
+                        It.IsAny<CancellationToken>()))
+                    .ReturnsAsync(new GetAuthenticationCredentialsResponse(expectations.AuthenticationUsername, expectations.AuthenticationPassword, message: null, authenticationTypes: null, responseCode: MessageResponseCode.NotFound));
             }
 
             PluginManager = new PluginManager(

--- a/test/NuGet.Core.Tests/NuGet.Credentials.Test/SecurePluginCredentialProviderTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.Credentials.Test/SecurePluginCredentialProviderTests.cs
@@ -361,6 +361,40 @@ namespace NuGet.Credentials.Test
             logger.Verify();
         }
 
+        [Fact]
+        public async Task GetAsync_WhenCredentialPluginIsUnableToAcquireCredentials_ReturnsNotFoundAsync()
+        {
+            var expectation = new TestExpectation(
+                operationClaims: new[] { OperationClaim.Authentication },
+                connectionOptions: ConnectionOptions.CreateDefault(),
+                pluginVersion: ProtocolConstants.CurrentVersion,
+                uri: _uri,
+                authenticationUsername: null,
+                authenticationPassword: null,
+                success: true,
+                messageCodeNotFound: true);
+
+            using (var test = new PluginManagerMock(
+                pluginFilePath: "a",
+                pluginFileState: PluginFileState.Valid,
+                expectations: expectation))
+            {
+                var discoveryResult = new PluginDiscoveryResult(new PluginFile("a", new Lazy<PluginFileState>(() => PluginFileState.Valid)));
+                var provider = new SecurePluginCredentialProvider(test.PluginManager, discoveryResult, canShowDialog: true, logger: NullLogger.Instance);
+
+                IWebProxy proxy = null;
+                var credType = CredentialRequestType.Unauthorized;
+                var message = "nothing";
+                var isRetry = false;
+                var isInteractive = false;
+                var token = CancellationToken.None;
+                var credentialResponse = await provider.GetAsync(_uri, proxy, credType, message, isRetry, isInteractive, token);
+
+                Assert.True(credentialResponse.Status == CredentialStatus.UserCanceled);
+                Assert.Null(credentialResponse.Credentials);                
+            }
+        }
+
         private PluginDiscoveryResult CreatePluginDiscoveryResult(PluginFileState pluginState = PluginFileState.Valid)
         {
             return new PluginDiscoveryResult(new PluginFile(Path.Combine(_testDirectory.Path, "plugin.exe"), new Lazy<PluginFileState>(() => pluginState)));

--- a/test/NuGet.Core.Tests/NuGet.Credentials.Test/SecurePluginCredentialProviderTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.Credentials.Test/SecurePluginCredentialProviderTests.cs
@@ -391,7 +391,7 @@ namespace NuGet.Credentials.Test
                 var credentialResponse = await provider.GetAsync(_uri, proxy, credType, message, isRetry, isInteractive, token);
 
                 Assert.True(credentialResponse.Status == CredentialStatus.UserCanceled);
-                Assert.Null(credentialResponse.Credentials);                
+                Assert.Null(credentialResponse.Credentials);
             }
         }
 

--- a/test/NuGet.Core.Tests/NuGet.Protocol.Tests/HttpSource/HttpSourceAuthenticationHandlerTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.Protocol.Tests/HttpSource/HttpSourceAuthenticationHandlerTests.cs
@@ -353,6 +353,55 @@ namespace NuGet.Protocol.Tests
         }
 
         [Fact]
+        public async Task SendAsync_WhenCredentialServiceReturnsNullThenTheHandlerAllowsRetriesUpToMaxAuthRetriesValue_Returns401Async()
+        {
+            // Arrange
+            var packageSource = new PackageSource("http://package.source.test");
+            var clientHandler = new HttpClientHandler();
+
+            var credentialService = Mock.Of<ICredentialService>();
+            
+            int retryCount = 0;
+            var innerHandler = new LambdaMessageHandler(
+                _ =>
+                {
+                    retryCount++;
+                    return new HttpResponseMessage(HttpStatusCode.Unauthorized);
+                });
+
+            var handler = new HttpSourceAuthenticationHandler(packageSource, clientHandler, credentialService)
+            {
+                InnerHandler = innerHandler
+            };
+
+            using (var client = new HttpClient(handler))
+            {
+                for (int i = 0; i < HttpSourceAuthenticationHandler.MaxAuthRetries + 1; i++)
+                {
+                    // Act
+                    var response = await client.SendAsync(request: new HttpRequestMessage(HttpMethod.Get, "http://foo"), CancellationToken.None);
+
+                    // Assert
+                    Assert.NotNull(response);
+                    Assert.Equal(HttpStatusCode.Unauthorized, response.StatusCode);
+                }                
+            }            
+
+            // Assert
+            Assert.Equal(HttpSourceAuthenticationHandler.MaxAuthRetries + 1, retryCount);
+
+            Mock.Get(credentialService)
+                .Verify(
+                    x => x.GetCredentialsAsync(
+                        packageSource.SourceUri,
+                        It.IsAny<IWebProxy>(),
+                        CredentialRequestType.Unauthorized,
+                        It.IsAny<string>(),
+                        It.IsAny<CancellationToken>()),
+                    Times.Exactly(HttpSourceAuthenticationHandler.MaxAuthRetries));
+        }
+
+        [Fact]
         public async Task SendAsync_WhenCredentialServiceThrows_Returns401()
         {
             // Arrange

--- a/test/NuGet.Core.Tests/NuGet.Protocol.Tests/HttpSource/HttpSourceAuthenticationHandlerTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.Protocol.Tests/HttpSource/HttpSourceAuthenticationHandlerTests.cs
@@ -360,7 +360,7 @@ namespace NuGet.Protocol.Tests
             var clientHandler = new HttpClientHandler();
 
             var credentialService = Mock.Of<ICredentialService>();
-            
+
             int retryCount = 0;
             var innerHandler = new LambdaMessageHandler(
                 _ =>
@@ -384,8 +384,8 @@ namespace NuGet.Protocol.Tests
                     // Assert
                     Assert.NotNull(response);
                     Assert.Equal(HttpStatusCode.Unauthorized, response.StatusCode);
-                }                
-            }            
+                }
+            }
 
             // Assert
             Assert.Equal(HttpSourceAuthenticationHandler.MaxAuthRetries + 1, retryCount);

